sparse/coo/core.py

```diff
@@ -1572,28 +1560,294 @@ def _keepdims(original, new, axis):
     return new.reshape(shape)
 
 
-def _mask(coords, idx, shape):
-    if isinstance(idx, numbers.Integral):
-        return coords == idx
-    elif isinstance(idx, slice):
-        step = idx.step if idx.step is not None else 1
-        if step > 0:
-            start = idx.start if idx.start is not None else 0
-            stop = idx.stop if idx.stop is not None else shape
-            return (coords >= start) & (coords < stop) & \
-                   (coords % step == start % step)
+def _mask(coords, indices, shape):
+    indices = _prune_indices(indices, shape)
+
+    ind_ar = np.empty((len(indices), 3), dtype=np.intp)
+
+    for i, idx in enumerate(indices):
+        if isinstance(idx, slice):
+            ind_ar[i] = [idx.start, idx.stop, idx.step]
         else:
-            start = idx.start if idx.start is not None else (shape - 1)
-            stop = idx.stop if idx.stop is not None else -1
-            return (coords <= start) & (coords > stop) & \
-                   (coords % step == start % step)
-    elif isinstance(idx, Iterable):
-        mask = np.zeros(len(coords), dtype=np.bool)
-        for item in idx:
-            mask |= _mask(coords, item, shape)
+            ind_ar[i] = [idx, idx + 1, 1]
+
+    mask, is_slice = _get_mask(coords, ind_ar)
+
+    if is_slice:
+        return slice(mask[0], mask[1], 1)
+    else:
         return mask
 
 
+def _prune_indices(indices, shape, prune_none=True):
+    """
+    Gets rid of the indices that do not contribute to the
+    overall mask, e.g. None and full slices.
+
+    Parameters
+    ----------
+    indices : tuple
+        The indices to the array.
+    shape : tuple[int]
+        The shape of the array.
+
+    Returns
+    -------
+    indices : tuple
+        The filtered indices.
+
+    Examples
+    --------
+    >>> _prune_indices((None, 5), (10,)) # None won't affect the mask
+    [5]
+    >>> _prune_indices((slice(0, 10, 1),), (10,)) # Full slices don't affect the mask
+    []
+    """
+    if prune_none:
+        indices = [idx for idx in indices if idx is not None]
+
+    i = 0
+    for idx, l in zip(indices[::-1], shape[::-1]):
+        if not isinstance(idx, slice):
+            break
+
+        if idx.start == 0 and idx.stop == l and idx.step == 1:
+            i += 1
+            continue
+
+        if idx.start == l - 1 and idx.stop == -1 and idx.step == -1:
+            i += 1
+            continue
+
+        break
+    if i != 0:
+        indices = indices[:-i]
+    return indices
+
+
+@numba.jit(nopython=True)
+def _get_mask(coords, indices):  # pragma: no cover
+    """
+    Gets the mask for the coords given the indices in slice format.
+
+    Parameters
+    ----------
+    coords : np.ndarray
+        The coordinates of the array.
+    indices : np.ndarray
+        The indices in the form of slices.
+
+    Returns
+    -------
+    mask : np.ndarray
+        The starts and stops in the mask.
```

Comment:
and steps?

Reply:
Steps turn out to always be 1 in this algorithm.